### ISSUE TITLE
AARCHXX: Potential priv TLS layout issue

### ISSUE THREAD

**Issue by ksco** [DynamoRIO/dynamorio#6225]
During the porting of RISCV64, I found an issue with the private TLS layout on the AARCHXX side. But I'm not sure about this, so here is an issue for further discussion.

Namely, in [loader_linux.c#L343](https://github.com/DynamoRIO/dynamorio/blob/master/core/unix/loader_linux.c#L343) and [loader_linux.c#L400](https://github.com/DynamoRIO/dynamorio/blob/master/core/unix/loader_linux.c#L400), dr_tp calculation seems incorrect.

Similar to RISC-V, ARM/Aarch64 uses the Variant I TLS layout, while x86 uses Variant II.

![Variant I](https://github.com/DynamoRIO/dynamorio/assets/15257961/03504329-03ce-4fe1-8c27-09037814073b)

![Variant II](https://github.com/DynamoRIO/dynamorio/assets/15257961/7d12e539-de7c-43e5-95b9-defca0b7c049)
> https://www.uclibc.org/docs/tls.pdf

So for AARCHXX, at line 343, `dr_tp` should be calculated as: `dr_tp = dr_tp + TLS_PRE_TCB_SIZE;`, at line 400, should be: `dr_tp = dr_tp - TLS_PRE_TCB_SIZE;`.

For RISC-V, refer to https://github.com/DynamoRIO/dynamorio/pull/6147, at line 343, `dr_tp` should be calculated as: `dr_tp = dr_tp + TLS_PRE_TCB_SIZE + sizeof(tcb_head_t);`, at line 400, should be: `dr_tp = dr_tp - TLS_PRE_TCB_SIZE - sizeof(tcb_head_t);`. The RISC-V part will be submitted as a new PR once this issue is confirmed.

**Comment by derekbruening** [comment]
What are the consequences of having it wrong?  Would only a private libpthread notice and that's why it's worked so far?

@AssadHashmi may know more here.

Xref ideas under #5437 on trying to not have to depend on libc/libpthread internal conventions but those are drastic changes.

**Comment by derekbruening** [comment]
Most of our own runs on very large apps are using statically linked clients + DR into the app, so while we haven't seen stability issues we aren't really using the private loader.

Pinging @AssadHashmi again to see if he has any more info.  @ksco if you're convinced we have it wrong you're probably right I would think.

**Comment by ksco** [comment]
> What are the consequences of having it wrong?

On RISC-V, without this patch, `drrun`ing the simplest assembly hello world would cause failure[1]. But on Aarch64, `privload_tls_init` is not even getting called when running `drrun echo hello`, so it worked fine without this change. I'm still getting familiar with the code base and client usage in general, and I don't understand what causes the difference between RISC-V and Aarch64. Isn't the `-private_loader` option enabled by default?

1.
```
<Application /media/debian/dynamorio/build/hello (2924).  Internal Error: DynamoRIO debug check failure: /media/debian/dynamorio/core/unix/tls_linux_risc.c:80 *get_dr_tls_base_addr() == NULL || *get_dr_tls_base_addr() == TLS_SLOT_VAL_EXITED
```

**Comment by ksco** [comment]
To be more specific, this https://github.com/ksco/dynamorio/commit/b6a0bb9c495ed84236daf868158d89910e05028e is the patch I would like to submit.

**Comment by ksco** [comment]
> Put a watchpoint in the debugger?

So, I set up a breakpoint at os.c:2286 (where `privload_tls_init` get called), here is log from Aarch64 and RISCV64:

> Aarch64
```
ksco@Ubuntu:/Users/ksco/Developer/dynamorio/build$ gdb bin64/drrun
GNU gdb (Ubuntu 13.1-2ubuntu2) 13.1
...
Reading symbols from bin64/drrun...
Reading symbols from /Users/ksco/Developer/dynamorio/build/bin64/drrun.debug...
(gdb) b os.c:2286
No source file named os.c.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (os.c:2286) pending.
(gdb) r echo hello
Starting program: /Users/ksco/Developer/dynamorio/build/bin64/drrun echo hello
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/aarch64-linux-gnu/libthread_db.so.1".
process 30620 is executing new program: /Users/ksco/Developer/dynamorio/build/lib64/debug/libdynamorio.so
<Starting application /usr/bin/echo (30620)>
<Initial options = -no_dynamic_options -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct >

Breakpoint 1, os_tls_app_seg_init (os_tls=0xfffdb3a6e000, segment=0xfffdb3a6e000) at /Users/ksco/Developer/dynamorio/core/unix/os.c:2286
2286	        os_tls->os_seg_info.priv_lib_tls_base = IF_UNIT_TEST_ELSE(
(gdb) p/x os_tls->app_lib_tls_base
$1 = 0x0
(gdb)
```

> RISCV64
```
debian@lpi4a:/media/debian/dynamorio/build$ gdb bin64/drrun
GNU gdb (Debian 13.2-1) 13.2
...
Reading symbols from bin64/drrun...
Reading symbols from /media/debian/dynamorio/build/bin64/drrun.debug...
(gdb) b os.c:2286
No source file named os.c.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (os.c:2286) pending.
(gdb) r hello
Starting program: /media/debian/dynamorio/build/bin64/drrun hello
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/riscv64-linux-gnu/libthread_db.so.1".
process 619 is executing new program: /media/debian/dynamorio/build/lib64/debug/libdynamorio.so
<Starting application /media/debian/dynamorio/build/hello (619)>
<Initial options = -no_dynamic_options -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct >

Breakpoint 1, os_tls_app_seg_init (os_tls=0x3df7d76000, segment=0x3df7d76000) at /media/debian/dynamorio/core/unix/os.c:2286
2286	        os_tls->os_seg_info.priv_lib_tls_base = IF_UNIT_TEST_ELSE(
(gdb) p/x os_tls->app_lib_tls_base
$1 = 0x3ff7fcf3a0
(gdb)
```

Idk what causes the difference, but maybe this is the reason why the Aarch64 port worked so far? The problematic path in `privload_tls_init` is never got executed.

**Comment by derekbruening** [comment]
I suggested a watchpoint, not a breakpoint: a watchpoint to see who wrote that non-zero value on RISCV.

**Comment by derekbruening** [comment]
Above you printed app_lib_tls_base -- but that shouldn't factor into the assert which is the slot where the DR TLS base is stored so I don't think they're related?

**Comment by ksco** [comment]
> I suggested a watchpoint, not a breakpoint: a watchpoint to see who wrote that non-zero value on RISCV that triggers the assert while it's zero on AArch64 (I assume it's zero; or is it the exited sentinel).

Under gdb, the error is gone away, the hello world program is executed normally. Sorry for not being explicit.
